Console command is not working with source_profile

Maybe because of an update AWS CLI or AWS APIs? I just started trying out Rain, so can't say how it worked "before".
`aws-vault login assumed-role` works fine, so probably should not be an AWS permission issue.

```
$ rain console -p assumed-role --debug

DEBUG: Generating sign-in URL
DEBUG: Loading AWS config
DEBUG: Reloading AWS credentials
panic: invalid character '<' looking for beginning of value [recovered]
	panic: invalid character '<' looking for beginning of value

goroutine 1 [running]:
github.com/aws-cloudformation/rain/internal/cmd.execute.func1()
	github.com/aws-cloudformation/rain/internal/cmd/wrap.go:77 +0x1a8
panic({0x106498280, 0x1400062ea38})
	runtime/panic.go:884 +0x204
github.com/aws-cloudformation/rain/internal/cmd/console.Open(0x0, {0x1058400dc, 0xe}, {0x0, 0x0}, {0x0, 0x0})
	github.com/aws-cloudformation/rain/internal/cmd/console/util.go:20 +0x2b0
github.com/aws-cloudformation/rain/internal/cmd/console.glob..func1(0x106926aa0?, {0x14000948cf0?, 0x3?, 0x3?})
	github.com/aws-cloudformation/rain/internal/cmd/console/console.go:30 +0x70
github.com/spf13/cobra.(*Command).execute(0x106926aa0, {0x14000948cc0, 0x3, 0x3})
	github.com/spf13/cobra@v1.4.0/command.go:860 +0x4ac
github.com/spf13/cobra.(*Command).ExecuteC(0x106928120)
	github.com/spf13/cobra@v1.4.0/command.go:974 +0x354
github.com/spf13/cobra.(*Command).Execute(...)
	github.com/spf13/cobra@v1.4.0/command.go:902
github.com/aws-cloudformation/rain/internal/cmd.execute(0x104f71594?)
	github.com/aws-cloudformation/rain/internal/cmd/wrap.go:86 +0x50
github.com/aws-cloudformation/rain/internal/cmd.Execute(0x6478f60?)
	github.com/aws-cloudformation/rain/internal/cmd/wrap.go:95 +0x1c
```

AWS config:
```
[profile iam-user-profile]
region = us-east-1
credential_process = aws-vault exec iam-user-profile --json

[profile assumed-role]
source_profile = iam-user-profile
region = us-east-1
role_arn = arn:aws:iam::xxxxxxxxxxxx:role/xxxx/xxxx
```

Versions:
```
$ rain --version
Rain v1.3.0 darwin/arm64
$ aws --version
aws-cli/2.10.3 Python/3.11.2 Darwin/21.6.0 source/arm64 prompt/off
```

Environment:
```
$ printenv | grep AWS
AWS_VAULT_KEYCHAIN_NAME=login
AWS_SDK_LOAD_CONFIG=1
```

--- a/internal/aws/aws.go
+++ b/internal/aws/aws.go
@@ -80,15 +80,15 @@ func loadConfig(ctx context.Context, sessionName string) *aws.Config {
 		panic(errors.New("unable to find valid credentials"))
 	}
 
+	if cfg.Region == "" {
+		panic(errors.New("a region was not specified. You can run 'aws configure' or choose a profile with a region"))
+	}
+
 	// Check for validity
 	creds, err = cfg.Credentials.Retrieve(context.Background())
 	if err != nil {
 		config.Debugf("Error retreiving creds: %s", err.Error())
 		panic(errors.New("could not establish AWS credentials; please run 'aws configure' or choose a profile"))
-	}
-
-	if cfg.Region == "" {
-		panic(errors.New("a region was not specified. You can run 'aws configure' or choose a profile with a region"))
 	}
 
 	return &cfg

--- a/internal/aws/console/console.go
+++ b/internal/aws/console/console.go
@@ -13,6 +13,7 @@ import (
 	"github.com/aws-cloudformation/rain/internal/aws"
 	"github.com/aws-cloudformation/rain/internal/aws/cfn"
 	"github.com/aws-cloudformation/rain/internal/aws/sts"
+	"github.com/aws-cloudformation/rain/internal/config"
 	"github.com/aws/smithy-go/ptr"
 )
 
@@ -39,25 +40,51 @@ func buildSessionString(sessionName string) (string, error) {
 		sessionName = nameParts[1]
 	}
 
+	config.Debugf("sessionName: %v", sessionName)
+
 	creds, err := aws.NamedConfig(sessionName).Credentials.Retrieve(context.Background())
 	if err != nil {
 		return "", err
 	}
 
-	return url.QueryEscape(fmt.Sprintf(`{"sessionId": "%s", "sessionKey": "%s", "sessionToken": "%s"}`,
-		creds.AccessKeyID,
-		creds.SecretAccessKey,
-		creds.SessionToken,
-	)), nil
+	unescaped := fmt.Sprintf(`{"sessionId": "%s", "sessionKey": "%s", "sessionToken": "%s"}`,
+		creds.AccessKeyID, creds.SecretAccessKey, creds.SessionToken)
+
+	config.Debugf("unescaped session string: %v", unescaped)
+
+	return url.QueryEscape(unescaped), nil
 }
 
 func getSigninToken(userName string) (string, error) {
 	sessionString, err := buildSessionString(userName)
 	if err != nil {
+		config.Debugf("buildSessionString failed")
 		return "", err
 	}
+	config.Debugf("sessionString: %v", sessionString)
 
-	resp, err := http.Get(fmt.Sprintf("%s?Action=getSigninToken&Session=%s&SessionDuration=%d", signinURI, sessionString, sessionDuration))
+	// Broken with source_profile and a role arn in .aws/config
+	//uri := fmt.Sprintf("%s?Action=getSigninToken&Session=%s&SessionDuration=%d",
+	//	signinURI, sessionString, sessionDuration)
+
+	// Try it without session duration (console sessions will be limited to 1 hour)
+	uri := fmt.Sprintf("%s?Action=getSigninToken&Session=%s",
+		signinURI, sessionString)
+
+	// Looks like this is the problem. SessionDuration is only valid when AssumeRole
+	// is called, so when source_profile is used, it must cause a call to
+	// GetFederationToken, which would require the use of DurationSeconds.
+
+	config.Debugf("uri: %v", uri)
+
+	// This page provides a good explanation of what we're doing here:
+	// https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_enable-console-custom-url.html
+
+	resp, err := http.Get(uri)
+	config.Debugf("resp.StatusCode: %v", resp.StatusCode)
+	if resp.StatusCode >= 300 && err == nil {
+		err = fmt.Errorf("Call to signin.aws.amazon.com resulted in a %v: %v", resp.StatusCode, resp.Status)
+	}
 	if err != nil {
 		return "", err
 	}
@@ -67,6 +94,8 @@ func getSigninToken(userName string) (string, error) {
 	if err != nil {
 		return "", err
 	}
+
+	config.Debugf("body: %s", body)
 
 	var out map[string]string
 	err = json.Unmarshal(body, &out)
@@ -84,10 +113,16 @@ func getSigninToken(userName string) (string, error) {
 
 // GetURI returns a sign-in uri for the current credentials and region
 func GetURI(service, stackName, userName string) (string, error) {
+
+	config.Debugf("GetURI %v, %v, %v", service, stackName, userName)
+
 	token, err := getSigninToken(userName)
 	if err != nil {
+		config.Debugf("getSigninToken failed")
 		return "", err
 	}
+
+	config.Debugf("token: %v", token)
 
 	if service == "" {
 		service = defaultService

--- a/internal/aws/console/console.go
+++ b/internal/aws/console/console.go
@@ -21,8 +21,7 @@ const signinURI = "https://signin.aws.amazon.com/federation"
 const issuer = "https://aws-cloudformation.github.io/rain/rain_console.html"
 const consoleURI = "https://console.aws.amazon.com"
 const defaultService = "cloudformation"
-
-//const sessionDuration = 43200
+const sessionDuration = 43200
 
 func buildSessionString(sessionName string) (string, error) {
 	if sessionName == "" {
@@ -65,26 +64,31 @@ func getSigninToken(userName string) (string, error) {
 	config.Debugf("sessionString: %v", sessionString)
 
 	// Broken with source_profile and a role arn in .aws/config
-	//uri := fmt.Sprintf("%s?Action=getSigninToken&Session=%s&SessionDuration=%d",
-	//	signinURI, sessionString, sessionDuration)
+	uriWithSessionDuration := fmt.Sprintf("%s?Action=getSigninToken&Session=%s&SessionDuration=%d",
+		signinURI, sessionString, sessionDuration)
 
 	// Try it without session duration (console sessions will be limited to 1 hour)
-	uri := fmt.Sprintf("%s?Action=getSigninToken&Session=%s",
+	uriWithoutSessionDuration := fmt.Sprintf("%s?Action=getSigninToken&Session=%s",
 		signinURI, sessionString)
 
-	// Looks like this is the problem. SessionDuration is only valid when AssumeRole
+	// SessionDuration is only valid when AssumeRole
 	// is called, so when source_profile is used, it must cause a call to
 	// GetFederationToken, which would require the use of DurationSeconds.
-
-	config.Debugf("uri: %v", uri)
+	// It's not clear how we could predict reliably which one will be used,
+	// so we try both URIs and see which one works. Not ideal.
 
 	// This page provides a good explanation of what we're doing here:
 	// https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_enable-console-custom-url.html
 
-	resp, err := http.Get(uri)
+	resp, err := http.Get(uriWithSessionDuration)
 	config.Debugf("resp.StatusCode: %v", resp.StatusCode)
 	if resp.StatusCode >= 300 && err == nil {
-		err = fmt.Errorf("call to signin.aws.amazon.com resulted in a %v: %v", resp.StatusCode, resp.Status)
+		config.Debugf("Retrying without SessionDuration after call to signin.aws.amazon.com resulted in a %v: %v",
+			resp.StatusCode, resp.Status)
+		resp, err = http.Get(uriWithoutSessionDuration)
+		if resp.StatusCode >= 300 && err == nil {
+			err = fmt.Errorf("call to signin.aws.amazon.com resulted in a %v: %v", resp.StatusCode, resp.Status)
+		}
 	}
 	if err != nil {
 		return "", err

--- a/internal/aws/console/console.go
+++ b/internal/aws/console/console.go
@@ -21,7 +21,8 @@ const signinURI = "https://signin.aws.amazon.com/federation"
 const issuer = "https://aws-cloudformation.github.io/rain/rain_console.html"
 const consoleURI = "https://console.aws.amazon.com"
 const defaultService = "cloudformation"
-const sessionDuration = 43200
+
+//const sessionDuration = 43200
 
 func buildSessionString(sessionName string) (string, error) {
 	if sessionName == "" {
@@ -83,7 +84,7 @@ func getSigninToken(userName string) (string, error) {
 	resp, err := http.Get(uri)
 	config.Debugf("resp.StatusCode: %v", resp.StatusCode)
 	if resp.StatusCode >= 300 && err == nil {
-		err = fmt.Errorf("Call to signin.aws.amazon.com resulted in a %v: %v", resp.StatusCode, resp.Status)
+		err = fmt.Errorf("call to signin.aws.amazon.com resulted in a %v: %v", resp.StatusCode, resp.Status)
 	}
 	if err != nil {
 		return "", err

--- a/internal/cmd/console/util.go
+++ b/internal/cmd/console/util.go
@@ -12,7 +12,7 @@ import (
 
 // Open generates a sign-in URL to the AWS console with an optional service and resource
 // If printOnly is true, the URL is printed to the console
-// If printOnly is fale, Open attempts to call the OS's browser with the URL
+// If printOnly is false, Open attempts to call the OS's browser with the URL
 func Open(printOnly bool, service, resource, userName string) {
 	spinner.Push("Generating sign-in URL")
 	uri, err := console.GetURI(service, resource, userName)

--- a/internal/config/version.go
+++ b/internal/config/version.go
@@ -5,5 +5,5 @@ const (
 	NAME = "Rain"
 
 	// VERSION is the application's version string
-	VERSION = "v1.3.0"
+	VERSION = "v1.3.3"
 )

